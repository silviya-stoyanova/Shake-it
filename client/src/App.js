import React, { Component } from 'react';
import { BrowserRouter } from 'react-router-dom'
import { ToastContainer } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css';
import './static/css/site.css'
import './static/css/scrollbar.css'

import sessionManager from './components/utilities/session-util'
import observer from './components/utilities/observer'

import MyRoutes from './components/routes/router'
import Header from './components/common/header'
import Footer from './components/common/footer'

const defaultValue = { isLogged: false }
const { Consumer: UserInfoConsumer, Provider: UserInfoProvider } = React.createContext(defaultValue)

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isLogged: sessionManager.isUserLogged(),
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }

    observer.subscribe('userLogin', this.onUserLogin)
    observer.subscribe('userLogout', this.onUserLogout)
  }

  onUserLogin = () => {
    this.setState(prevState => ({ // asyncronous
      isLogged: true,
      username: sessionManager.getUsername(),
      role: sessionManager.getUserRole()
    }))
  }

  onUserLogout = () => {
    this.setState(prevState => ({
      isLogged: false,
      username: '',
      role: []
    }))
  }

  render() {
    const { isLogged, username, role } = this.state

    // console.log(this.props.navigation)

    return (
      <UserInfoProvider value={{ isLogged, username, role }}>
        <BrowserRouter>
          <ToastContainer />
          <Header />

          {
            // <div className="main-container">
          }
          <MyRoutes />
          {
            // </div>
          }

          <Footer />
        </BrowserRouter>
      </UserInfoProvider>
    )
  }

  componentDidMount() {
    sessionManager.isUserLogged()
      ? this.setState({ isLogged: true })
      : this.setState({ isLogged: false })
  }
}

export { UserInfoConsumer }
export default App