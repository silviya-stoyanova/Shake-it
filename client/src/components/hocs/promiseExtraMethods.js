import requester from '../../utilities/requests-util'
import observer from '../../utilities/observer'
import sessionManager from '../../utilities/session-util'

const promiseExtraMethods = {
    user: {
        onLoginSuccess(res) {
            sessionManager.saveSession(res.authtoken, res.username, res.role)
            observer.trigger('userLogin')
        },
        onRegisterSuccess(res, data) {
            let nextPromise = requester.login(data)
            this.handleFetchPromise(nextPromise, this.onLoginSuccess, this.onLoginFail, data) // { username: data.username, password: data.password }
        },
        onLoginFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'correct',
                    passClass: 'error',
                    repeatPassClass: 'error'
                }
            }))
        },
        onRegisterFail(err) {
            this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    userClass: 'error',
                    passClass: 'correct',
                    repeatPassClass: 'correct'
                }
            }))
        }
    },

    product: {
        onProductPromiseSuccess(err) {
            return err // do nothing
        },

        onProductPromiseFail(err) {
            if (!err.titleIsTaken) {
                this.props.history.push('/')
            }
        }
    }
}

export default promiseExtraMethods