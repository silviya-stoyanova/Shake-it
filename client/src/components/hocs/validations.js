import { toast } from 'react-toastify'

const userValidations = {
    validateUserData(formType, username, password, repeatPassword, updateState) {
        let [userClass, passClass, repeatPassClass] = ['error', 'error', 'error']
        let result = {
            username: 'invalid',
            password: 'invalid',
            repeatPassword: 'invalid'
        }

        if (username && username.trim().length >= 5) {
            userClass = 'correct'
            result.username = 'valid'
        }
        if (password && password.length >= 8) {
            passClass = 'correct'
            result.password = 'valid'
        }

        if (repeatPassword) {
            if (repeatPassword && repeatPassword.length >= 8) {
                repeatPassClass = 'correct'
                result.repeatPassword = 'valid'
            }

            if (password !== repeatPassword) {
                [passClass, repeatPassClass] = ['error', 'error']
                result.password = 'invalid'
                result.repeatPassword = 'invalid'
            }
        }

        updateState && this.setState(prevState => ({
            userInfo: {
                ...prevState.userInfo,
                userClass,
                passClass,
                repeatPassClass
            }
        }))

        return result
    },

    validateUserOnSubmit(formType, username, password, repeatPassword) {
        const isInputValid = this.validateUserData(formType, username, password, repeatPassword, false)
        let isValid = false
        let errorMsg = ''

        if (formType === 'login') {
            if (isInputValid.username === 'invalid' || isInputValid.password === 'invalid') {
                errorMsg = 'Either your username or password is invalid.'
            }
        } else if (formType === 'register') {
            if (isInputValid.password === 'invalid' || isInputValid.repeatPassword === 'invalid') {
                errorMsg = 'Both passwords must consist of at least 8 characters and they both must match!'
            }
            if (isInputValid.username === 'invalid') {
                errorMsg = 'Please enter a valid username!'
            }
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast'
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

const productValidations = {
    validateProductData(formType, title, description, image, price, updateState) {
        let [titleClass, descriptionClass, priceClass] = ['error', 'error', 'error', 'error']
        let result = {
            title: 'invalid',
            description: 'invalid',
            image: 'invalid',
            price: 'invalid',
        }

        if (title && title.length > 2 && title.length <= 50) {
            titleClass = 'correct'
            result.title = 'valid'
        }
        if (description && description.length > 9 && description.length <= 250) {
            descriptionClass = 'correct'
            result.description = 'valid'
        }

        // validate only when creating a new product
        if (formType === 'create' && image) {
            result.image = 'valid'
        }
        if (price && Number(price) > 0) {
            priceClass = 'correct'
            result.price = 'valid'
        }

        updateState && this.setState(prevState => ({
            productInfo: {
                ...prevState.productInfo,
                titleClass,
                descriptionClass,
                priceClass
            }
        }))
        return result
    },

    validateProductOnSubmit(formType, title, description, image, price) {
        const updateState = false
        const isInputValid = this.validateProductData(formType, title, description, image, price, updateState)
        let isValid = false
        let errorMsg = ''

        if (isInputValid.price === 'invalid') {
            errorMsg = 'Please enter a valid price.'
        }
        // when adding a product
        if (formType === 'create' && isInputValid.image === 'invalid') {
            errorMsg = 'Please provide an image.'
        }
        if (isInputValid.description === 'invalid') {
            errorMsg = 'The description must be at least 10 symbols long.'
        }
        if (isInputValid.title === 'invalid') {
            errorMsg = 'The title length must be between 3 and 50 symbols including.'
        }
        if (errorMsg) {
            toast.info(errorMsg, {
                className: 'error-toast',
            })
            return isValid
        }

        isValid = true
        return isValid
    }
}

export { userValidations }
export { productValidations }