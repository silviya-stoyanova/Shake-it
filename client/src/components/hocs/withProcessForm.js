import React, { Component } from 'react'
import { toast } from 'react-toastify'
import requester from '../../utilities/requests-util'
import sessionManager from '../../utilities/session-util'

const withProcessForm = (Form, formType, validations, initialData, requestType, promiseExtraMethods, testServiceOnMount, testServiceOnSubmit) => {
    return class extends Component {
        constructor(props) {
            super(props)
            this.state = {
                data: initialData
            }

            this.validateData = validations ? validations.validateData.bind(this) : null
            this.validateOnSubmit = validations ? validations.validateOnSubmit.bind(this) : null

            this.promiseSuccess = promiseExtraMethods ? promiseExtraMethods.success.bind(this) : null
            this.promiseFail = promiseExtraMethods ? promiseExtraMethods.fail.bind(this) : null

            this.handleInputChange = this.handleInputChange.bind(this)
            this.handleFetchPromise = this.handleFetchPromise.bind(this)
            this.handleFormSubmit = this.handleFormSubmit.bind(this)
        }

        // static defaultProps = {
        //     service: requester.getProductInfo
        // }

        async handleInputChange({ target }) {
            console.log(target.files ? target.files[0] : null)

            await this.setState(prevState => ({
                data: {
                    ...prevState.data,
                    [target.name]: target.files ? target.files : target.value, // .trim()
                    uploadedImg: target.files ? URL.createObjectURL(target.files[0]) : prevState.data.uploadedImg
                    // uploadedImg: target.files ? target.files[0] : prevState.data.uploadedImg
                }
            }))

            const updateState = true
            const { data } = this.state

            // console.log(data.uploadedImg)

            this.validateData(formType, data, updateState)
        }

        handleFetchPromise(promise, onSuccessFunc, onFailFunc, data = '') {
            promise.then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()

            }).then(res => {
                toast.info(res.success, {
                    className: 'success-toast'
                })

                // data.redirectOnSuccess &&
                this.props.history.push('/')

                onSuccessFunc && onSuccessFunc(res, data)

            }).catch(err => {
                err.json().then(error => {
                    toast.info(error.message, {
                        className: 'error-toast'
                    })

                    onFailFunc && onFailFunc(error)
                })
            })
        }

        async handleFormSubmit(event) {
            // console.log(event.preventDefault())
            
            event.preventDefault()
            const { data } = this.state
            const jwtToken = sessionManager.getUserInfo().authtoken
            let isValid = this.validateOnSubmit(formType, data)

            if (!isValid) {
                return
            }

            if (testServiceOnSubmit) {
                testServiceOnSubmit()
                    // to see the message inside the response, uncomment the next two commented out lines
                    // .then(res => res.json())
                    .then(res => {
                        // console.log(res.success)
                        this.props.history.push('/')
                    })
                    .catch(err => {
                        return err
                    })

            } else {
                const promise = requester[requestType](data, jwtToken)
                this.handleFetchPromise(promise, this.promiseSuccess, this.promiseFail, data)
            }
        }

        render() {
            return <Form {...this.state} handleInputChange={this.handleInputChange} handleFormSubmit={this.handleFormSubmit} />
        }

        componentDidMount() {
            if (formType === 'create' || formType === 'login' || formType === 'register') {
                return
            }
            // let promise = requester.getProductInfo(this.state.productInfo._id)
            // this.handleFetchPromise(promise, null, null, { redirectOnFail: true })
            // requester.getProductInfo(this.state.data._id)

            const serviceReq = testServiceOnMount || requester.getProductInfo  // new

            const productId = this.props.match                      // new
                ? this.props.match.params.productId
                : null
            // const productId = this.props.match.params.productId

            // requester.getProductInfo(productId)                  // prev
            serviceReq && serviceReq(productId)                     // new
                .then(res => {
                    if (!res.ok) {
                        return Promise.reject(res)
                    }
                    return res.json()
                })
                .then(async res => {
                    await this.setState({
                        data: res
                    })
                })
                .catch(error => {
                    this.props.history.push('/')

                    error.json().then(err => {
                        toast.info(err.message, {
                            className: 'error-toast'
                        })
                    })
                })
        }
    }
}

export default withProcessForm