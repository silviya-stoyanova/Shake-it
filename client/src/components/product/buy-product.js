import React from 'react'

const BuyProduct = () => {
    return {




    }
}

export default BuyProduct