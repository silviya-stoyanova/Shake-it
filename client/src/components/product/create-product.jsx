import React from 'react'
import withProcessProductForm from './with-product-forms-hoc'
import useTitle from '../page-title/useTitle'

const CreateProduct = (props) => {
    useTitle('Add a new product')
    const { titleClass, descriptionClass, priceClass, uploadedImg, handleInputChange, handleFormSubmit } = props

    return (
        <div className="form" >
            <form onSubmit={handleFormSubmit} encType="multipart/form-data" >
                <div className="form-type">Add a product</div>
                <hr />

                <div className="form-fields-wrapper">
                    <label htmlFor="title">Title</label>
                    <input autoFocus onChange={handleInputChange} type="text" name="title" className={titleClass} id="title" />

                    <label htmlFor="description">Description</label>
                    <textarea type="text" onChange={handleInputChange} name="description" className={descriptionClass} id="description"></textarea>

                    <div className="update-img-container">
                        <label className="input-file-container" htmlFor="image">
                            <div className="upload-img-text-wrapper">
                                <span className="upload-img-text">Upload an image</span>
                                <img className="update-img" style={{ left: '20%', top: '5%' }} src={require('../../static/images/update-pic.png')} alt="update-product" />
                            </div>

                            <div className="img-visualiser">
                                {uploadedImg
                                    ? <img src={uploadedImg} className="add-product-img" alt="product-img" />
                                    : <img src={require('../../static/images/create-product-img.png')} className="add-product-img opacity-img" alt="product-img" />
                                }
                            </div>
                        </label>
                        <input type="file" accept="image/*" onChange={handleInputChange} name="image" id="image" />
                    </div>

                    {/* <div>
                            <label className="update-img-content" htmlFor="profilePic">
                                <img className="update-img" src="" alt="update-your-profile-picture" />
                            </label>
                            <input onChange={handleInputChange} type="file" name="profilePic" id="profilePic" />
                        </div> */}


                    <label htmlFor="price">Price</label>
                    <input type="number" onChange={handleInputChange} name="price" className={priceClass} id="price" min="0" step="0.01" />
                </div>

                <hr />
                <button className="button" type="submit">Submit</button>
            </form>
        </div>
    )
}

// export default CreateProduct
export default withProcessProductForm(CreateProduct, 'create')