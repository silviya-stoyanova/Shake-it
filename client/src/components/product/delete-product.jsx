import React from 'react'
import { Link } from 'react-router-dom'
import '../../static/css/products.css'
import withProcessProductForm from './with-product-forms-hoc'
import useTitle from '../page-title/useTitle'

const DeleteProduct = (props) => {
    const { image, title, description, price, handleFormSubmit } = props
    useTitle(`Delete ${title}`)

    return (
        <div className="form" >
            <form onSubmit={handleFormSubmit}>
                <div className="form-type">Are you sure you want</div>
                <div className="form-type">to delete this product?</div>
                <hr />

                {image
                    ? <div className="form-fields-wrapper">
                        <div className="product-title">{title}</div>

                        <img src={'data:image/png;base64, ' + image} alt={title} className="product-img" />

                        <label htmlFor="description">Description</label>
                        <textarea defaultValue={description} type="text" id="description" disabled></textarea>

                        <div className="price">
                            <label htmlFor="price">Price:</label>
                            <input value={price} id="price" disabled />
                            <span className="price-sign">$</span>
                        </div>
                    </div>

                    : <img src={require('../../static/images/loading-circle.gif')} alt={'loading'} className="product-img" />
                }

                <hr />
                <button className="button btn-del" type="submit">Yes</button>
                <Link to="/" className="button btn-del">No</Link>
            </form>
        </div>
    )
}

export default withProcessProductForm(DeleteProduct, 'delete')