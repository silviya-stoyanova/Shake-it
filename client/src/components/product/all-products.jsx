import React, { Component, Fragment } from 'react'
import { Link } from 'react-router-dom'
import { toast } from 'react-toastify'
import { UserInfoConsumer } from '../../App'
import requester from '../../utilities/requests-util'
import '../../static/css/products.css'

class AllProducts extends Component {
    constructor(props) {
        super(props)
        this.state = {
            products: [],
            isFetched: false
        }

        this.renderAllProducts = this.renderAllProducts.bind(this)
    }

    userFunctionalities = (data, p) => {
        return data.isLogged
            && <Fragment>
                <Link
                    to={{
                        pathname: `/product/like/${p._id}`
                    }} className="product-actions-btn">like
                </Link>
                {
                    data.role === 'Admin'
                    && <Fragment>
                        <Link
                            to={{
                                pathname: `/product/edit/${p._id}`
                            }} className="product-actions-btn">edit
                            </Link>

                        <Link
                            to={{
                                pathname: `/product/delete/${p._id}`
                            }} className="product-actions-btn">delete
                            </Link>
                    </Fragment>
                }
                <Link
                    to={{
                        pathname: `/cart/add/${p._id}`
                    }} className="product-actions-btn add-to-cart-btn">add to cart
                </Link>
            </Fragment>
    }

    renderAllProducts(data) {
        const { products } = this.state

        return (
            <Fragment>
                {
                    products.map(p => {
                        return (
                            <div key={p._id} className="product-wrapper" >
                                <Link className="product-title" to={{
                                    pathname: `/product/details/${p._id}`
                                }}>
                                    {p.title}
                                </Link>

                                <Link to={{
                                    pathname: `/product/details/${p._id}`
                                }} ><img src={'data:image/png;base64, ' + p.image} alt={p.title} className="product-img" />
                                </Link>

                                <div className="product-actions">
                                    <span className="product-price">Price: {p.price}<span className="price-sign">$</span></span>
                                    <span className="product-likes">{p.likes ? p.likes.length : 0} â™¥</span>
                                    {this.userFunctionalities(data, p)}
                                </div>
                            </div>
                        )
                    })
                }
            </Fragment >
        )
    }

    // <h1>Our site is under constructtion. We will be ready soon..</h1>
    // <h2>xoxo</h2>
    render() {
        const { products, isFetched } = this.state

        return (
            <div className='content-wrapper' >

                <UserInfoConsumer>
                    {(data) => (

                        isFetched
                            ? products.length
                                ? this.renderAllProducts(data)
                                : (<div className="no-products-container">
                                    <div className="no-products-text-container">
                                        <h1>Our products are so desired that we could not predict we would run out of stock this soon..</h1>
                                        <h2>Sorry!</h2>
                                    </div>

                                    <div className="no-products-img"></div>
                                </div>)
                            : < div className="product-wrapper" >
                                <img src={require('../../static/images/loading-circle.gif')} alt="loading-img" className="product-img" />
                            </ div>
                    )}
                </UserInfoConsumer>
            </div>
        )
    }

    componentDidMount() {
        document.title = 'Shake it - Your shake is a click away'

        requester.getAllProducts()
            .then(res => {
                if (!res.ok) {
                    return Promise.reject(res)
                }
                return res.json()
            })
            .then(allProducts => {
                this.setState({
                    products: allProducts,
                    isFetched: true
                })
            })
            .catch(err => {
                err.json()
                    .then(res => {
                        toast.info(res.message, {
                            className: 'error-toast',
                        })
                    })
            })
    }
}

export default AllProducts