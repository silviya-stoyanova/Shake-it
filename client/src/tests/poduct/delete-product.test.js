//!! https://stackoverflow.com/questions/50682726/how-to-unit-test-a-react-event-handler-that-contains-history-push-using-jest-and
// what does jest.fn() ?
// https://airbnb.io/enzyme/docs/api/ShallowWrapper/simulate.html
// https://stackoverflow.com/questions/44419318/how-to-test-form-submission-in-react-with-jest-and-enzyme-cannot-read-property

import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import { UserInfoProvider } from '../../App'
import withProcessForm from '../../components/hocs/withProcessForm'
import { DeleteProduct } from '../../components/product/delete-product'
import { productValidations } from '../../components/hocs/validations'
import AuthRoute from '../../components/routes/auth-route'

// when someone accesses non-existing product's details
// => just before they are redirected, they will see loading - circle.gif
// => and then will be redirected

// scenarios:
//
//// a [guest] accesses delete-product page => and then will be redirected => DeleteProduct component should not be rendered??
//
//// a [user] accesses delete-product page => and then will be redirected => nothing should be rendered??
//
// The [Admin] accesses delete-[existing]-product page
//              -- and the data is fetched
//              -- and the data is not fetched yet
// The [Admin] accesses delete-[non-existing]-product page      // stuff to happen here
//

const mockNonExistingProductFetch = () => {



}

const mockExistingProductFetch = () => {
    const sampleData = JSON.stringify(
        {
            _id: 1,
            title: 'Gotino zaglavie',
            image: 'дилишъс-шейк.jpg',
            description: 'Представям на Вашето внимание един уникален млечен шейк',
            price: 9.99,
            likes: ['firstUserId', 'secondUserId', 'thirdUserId', 'fourthuserId', 'fifthUserId', 'sixthUserId', 'seventhUserId', 'eighthUserId']
        })

    return Promise.resolve(new Response(sampleData))
}

const mockPoductDelete = () => {
    return Promise.resolve({ success: 'Product deleted successfully!' })
}

const initialData = {
    title: '', description: '', image: '', price: '',
}

const sampleEvent = {
    preventDefault: () => {
        console.log('I am fake prevent-defaulter.')
    }
}

const historyMock = {
    push: jest.fn()
}

describe('tests for the component DeleteProduct', () => {

    describe('tests for guest functionalities', () => {

        it('should return redirect to the home page when a [guest] accesses delete-product page', () => {
            const defaultUserValue = { isLogged: false, username: '', role: '' }

            const wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue}>
                        <AuthRoute path="/product/delete/:productId" component={DeleteProduct} role="Admin" isTest={true} />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.debug()).toMatchSnapshot()
            expect(wrapper.find('Redirect[to="/"]').length).toEqual(1)
        })
    })

    describe('tests for user functionalities', () => {

        it('should return redirect to the home page when a [user] accesses delete-product page', () => {
            const defaultUserValue = { isLogged: true, username: 'lalala', role: 'User' }

            const wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue}>
                        <AuthRoute path="/product/delete/:productId" component={DeleteProduct} role="Admin" isTest={true} />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            expect(wrapper.debug()).toMatchSnapshot()
            expect(wrapper.find('Redirect[to="/"]').length).toEqual(1)
        })
    })

    describe('tests for admin functionalities', () => {

        it("should return product's delete page with it's [fetched] data, when The [Admin] tries to delete [existing] product", () => {

            const TestHoc = withProcessForm(DeleteProduct, 'delete', productValidations, initialData, null, null, mockExistingProductFetch, mockPoductDelete)
            const defaultUserValue = { isLogged: true, username: 'admin', role: 'Admin' }

            const wrapper = mount(
                <BrowserRouter>
                    <UserInfoProvider value={defaultUserValue}>
                        <TestHoc history={historyMock} />
                    </UserInfoProvider>
                </BrowserRouter>
            )

            mockExistingProductFetch()
                .then(res => res.json())
                .then(res => {
                    wrapper.setState({
                        data: res
                    })

                    wrapper.update()

                    expect(wrapper.debug()).toMatchSnapshot()

                    expect(wrapper.find('div.product-title').length).toEqual(1)
                    expect(wrapper.find('div.product-title').text()).toEqual('Gotino zaglavie')

                    expect(wrapper.find('img.product-img').length).toEqual(1)
                    expect(wrapper.find('img.product-img[src="data:image/png;base64, дилишъс-шейк.jpg"]').length).toEqual(1)

                    expect(wrapper.find('textarea[id="description"]').length).toEqual(1)
                    expect(wrapper.find('textarea[id="description"]').text()).toEqual('Представям на Вашето внимание един уникален млечен шейк')

                    expect(wrapper.find('div.price').length).toEqual(1)
                    expect(wrapper.find('input[id="price"]').length).toEqual(1)
                    expect(wrapper.find('input[id="price"]').prop('value')).toEqual(9.99)

                    expect(wrapper.find('button.btn-del').length).toEqual(1)
                    expect(wrapper.find('a[href="/"]').hasClass('btn-del')).toEqual(true)

                    wrapper.find('form').simulate('submit', sampleEvent)
                    // expect(wrapper.find('form').length).toEqual(0)
                    // expect(wrapper.debug()).toMatchSnapshot()
                    wrapper.update()

                    // expect(historyMock.push).toHaveBeenLastCalledWith('/') 
                    try {
                        expect(historyMock.push).toBeCalled()
                    } catch (err) {
                        console.log(err)
                    }
                    console.log(historyMock.push.mock.calls)


                })

        })

    })
})