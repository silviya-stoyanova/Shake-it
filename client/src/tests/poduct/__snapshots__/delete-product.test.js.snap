// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for the component DeleteProduct tests for admin functionalities should return product's delete page with it's [fetched] data, when The [Admin] tries to delete [existing] product 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <Component history={{...}}>
      <DeleteProduct data={{...}} handleInputChange={[Function: bound handleInputChange]} handleFormSubmit={[Function: bound handleFormSubmit]}>
        <div className=\\"form\\">
          <form onSubmit={[Function: bound handleFormSubmit]}>
            <div className=\\"form-type\\">
              Are you sure you want
            </div>
            <div className=\\"form-type\\">
              to delete this product?
            </div>
            <hr />
            <div className=\\"form-fields-wrapper\\">
              <div className=\\"product-title\\">
                Gotino zaglavie
              </div>
              <img src=\\"data:image/png;base64, дилишъс-шейк.jpg\\" alt=\\"Gotino zaglavie\\" className=\\"product-img\\" />
              <label htmlFor=\\"description\\">
                Description
              </label>
              <textarea defaultValue=\\"Представям на Вашето внимание един уникален млечен шейк\\" type=\\"text\\" id=\\"description\\" disabled={true} />
              <div className=\\"price\\">
                <label htmlFor=\\"price\\">
                  Price:
                </label>
                <input value={9.99} id=\\"price\\" disabled={true} />
                <span className=\\"price-sign\\">
                  $
                </span>
              </div>
            </div>
            <hr />
            <button className=\\"button btn-del\\" type=\\"submit\\">
              Yes
            </button>
            <Link to=\\"/\\" className=\\"button btn-del\\">
              <LinkAnchor className=\\"button btn-del\\" href=\\"/\\" navigate={[Function: navigate]}>
                <a className=\\"button btn-del\\" href=\\"/\\" onClick={[Function: onClick]}>
                  No
                </a>
              </LinkAnchor>
            </Link>
          </form>
        </div>
      </DeleteProduct>
    </Component>
  </Router>
</BrowserRouter>"
`;

exports[`tests for the component DeleteProduct tests for guest functionalities should return redirect to the home page when a [guest] accesses delete-product page 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <AuthRoute path=\\"/product/delete/:productId\\" component={[Function: DeleteProduct]} role=\\"Admin\\" isTest={true}>
      <Redirect to=\\"/\\">
        <Lifecycle onMount={[Function: onMount]} onUpdate={[Function: onUpdate]} to=\\"/\\" />
      </Redirect>
    </AuthRoute>
  </Router>
</BrowserRouter>"
`;

exports[`tests for the component DeleteProduct tests for user functionalities should return redirect to the home page when a [user] accesses delete-product page 1`] = `
"<BrowserRouter>
  <Router history={{...}}>
    <AuthRoute path=\\"/product/delete/:productId\\" component={[Function: DeleteProduct]} role=\\"Admin\\" isTest={true}>
      <Redirect to=\\"/\\">
        <Lifecycle onMount={[Function: onMount]} onUpdate={[Function: onUpdate]} to=\\"/\\" />
      </Redirect>
    </AuthRoute>
  </Router>
</BrowserRouter>"
`;
